Address comments from Mark (#74)
ietf-wg-dnsop/draft-ietf-dnsop-structured-dns-error: draft-ietf-dnsop-structured-dns-error.md

3. The EXTRA-TEXT field MUST be an I-JSON message {{!RFC7493}}. If the client fails
   to parse the field as valid JSON, it MUST treat the data as invalid and
   MUST NOT process it according to this specification.  The client MAY process
   the EXTRA-TEXT field as unstructured text as specified in {{!RFC8914}}.

4. The DNS response MUST also contain an extended error code of
   "Blocked by Upstream DNS Server", "Blocked" or "Filtered" {{!RFC8914}}, otherwise
   the EXTRA-TEXT field is discarded.

5. If the JSON object contains an "s" field and the sub-error code
   is not defined as applicable to the accompanying Extended DNS Error
   (EDE) code, the client MUST ignore the value of the "s" field
   and continue processing the remaining fields in accordance with this
   specification.

6. If the EXTRA-TEXT field does not contain at least one of the JSON
   names "c", "j", or "s", or if all of the fields that are present have
   empty values, the entire JSON object MUST be discarded.

7. If the "c" field contains any contact URIs that use a scheme not registered
   in the {{IANA-Contact}} registry, only those URIs MUST be discarded. Contact

>    URIs using registered schemes can be processed.

Reviewer: ```suggestion
7. If a Contact URI in the "c" field uses a scheme not registered
in the {{IANA-Contact}} registry, those URIs are discarded. Contact
URIs using registered schemes can be processed.
```

Reviewer: Tried to simplify language on this one

  profile for DoT, the DNS client MAY process the EXTRA-TEXT field of the
  DNS response.

11. The DNS client MUST ignore any other JSON names that it does not support.

> Note that the strict and opportunistic privacy profiles as defined in {{!RFC8310}} only apply to DoT; there has been
no such distinction made for DoH.

## Structured DNS Error (SDE) EDNS(0) Option Format {#SDE}

The Structured DNS Error (SDE) EDNS(0) option is used by a client to
indicate support for I-JSON encoding in the EXTRA-TEXT field of an
Extended DNS Error (EDE) option.

> The SDE option has no OPTION-DATA. The OPTION-LENGTH field MUST

Reviewer: Shouldn't these two processing requirements be in Client Generating Request and Server Generating Response?

Author: The detailed steps to generate SDE are already discussed in Client Generating Request and Server Generating Response sections.

Reviewer: OPTION-LENGTH is not mentioned in the processing steps. It's just a style thing where to place it, so it's fine.

be set to 0. A server receiving an SDE option with a non-zero
OPTION-LENGTH MUST ignore the option.

The presence of the SDE option in a query indicates that the client
supports processing the EXTRA-TEXT field in accordance with this

> specification.

Reviewer: do we also want to say, "the server provides no indication it supports this specification, until and unless the server happens to filter a response and includes I-JSON", or something like that?

Bigger question: do we want the ability for the server to say "yes, I support it" in the protocol?

If not "in the protocol", perhaps it should be done with [DNS Resolver Information](https://www.rfc-editor.org/rfc/rfc9606.html).  In a separate I-D.

Author: > do we also want to say, "the server provides no indication it supports this specification, until and unless the server happens to filter a response and includes I-JSON", or something like that?

It is already discussed in the Server Generating Response section (2nd and 3rd paragraphs). please check. 

> 
> Bigger question: do we want the ability for the server to say "yes, I support it" in the protocol?
> 
> If not "in the protocol", perhaps it should be done with [DNS Resolver Information](https://www.rfc-editor.org/rfc/rfc9606.html). In a separate I-D.

RFC9606 advertises the possible EDE codes that can be returned by this DNS resolver. It does not signal whether structured error data is supported by the DNS server.

Reviewer: RFC9606 also has an infourl, which is not an EDE code returned by the DNS resolver. An additional key could be defined using RFC9606 which says "I support draft-ietf-dnsop-structured-dns-error"; I don't see anything that prohibits doing that, should we want/need it in the future.

New JSON names are registered via IETF Review ({{Section 4.8 of !RFC8126}}).

The "Mandatory" column is informational only. This specification does not define any mandatory JSON names.

> To preserve backward compatibility, any new JSON names registered after publication of this document MUST set the “Mandatory” column to “N”. Future extensions cannot introduce mandatory JSON attributes, as existing implementations are required to ignore unknown JSON names (see {{client-processing}}).

Reviewer: ```suggestion
As existing clients ignore unknown JSON names (see {{client-processing}}), new JSON names will likely need to set the Mandatory column to "N".
```

Do we want a way to version this spec, perhaps by allowing SDE length to be non-zero?  That is, this specification would be the "length=0" version of the protocol, and a subsequent version would have SDE length=1 and its version number as the SDE value.  Seems an opportune place to allow versioning.

I re-read how the SDE length=0 is spec'd, and a subsequent version can just say something like "SDE length=0 means the original version, and SDE length=1 value=2 means this new version 2".  So I think we're good and we can version just fine.

Author: I don't see a need to support extensibility. Can you describe the incompatible evolution scenario that cannot be handled via the existing extensibility mechanisms ?

We already have 
New JSON names (registry)
New sub-error codes (registry)
New EDE codes (registry)
